src/react/HomePage.js

```diff
@@ -29,185 +29,166 @@ import {
 
-export default class HomePage extends React.Component {
-  constructor(props) {
-    super(props);
-
-    this.astarte = this.props.astarte;
-
-    this.updateDeviceStats = this.updateDeviceStats.bind(this);
-    this.updateInterfaceNames = this.updateInterfaceNames.bind(this);
-    this.updateTriggerNames = this.updateTriggerNames.bind(this);
-    this.redirectToLastInterface = this.redirectToLastInterface.bind(this);
-    this.updateStatus = this.updateStatus.bind(this);
-    this.syncData = this.syncData.bind(this);
-
-    this.queryFlowHealth = this.astarte.config.enableFlowPreview;
-
-    this.state = {
-      connectedDevices: "loading",
-      totalDevices: "loading",
-      interfaces: "loading",
-      triggers: "loading",
-      appengineStatus: "loading",
-      realmManagementStatus: "loading",
-      pairingStatus: "loading",
-      flowStatus: this.queryFlowHealth ? "loading" : null
-    };
+const WaitForData = ({status, fallback, children}) => {
+  switch (status) {
+    case "ok":
+      return children;
 
-    this.syncData();
-  }
+    case "err":
+    case "loading":
+      return fallback || null;
 
-  componentDidMount() {
-    this.astarte.addListener('credentialsChange', this.syncData)
+    default:
+      return null;
   }
+};
 
-  componentWillUnmount() {
-    this.astarte.removeListener('credentialsChange', this.syncData)
-  }
+const useAstarteHealth = (queryAstarteHealth) => {
+  const [status, setStatus] = useState("loading");
 
-  syncData() {
-    this.astarte.getDevicesStats()
-      .then(this.updateDeviceStats)
-      .catch((e) => this.handleSectionError("connectedDevices", e));
+  useEffect(() => {
+    if (!queryAstarteHealth) {
+      setStatus("err");
+      return;
+    }
+
+    let ignore = false;
 
-    this.astarte.getInterfaceNames()
-      .then(this.updateInterfaceNames)
-      .catch((e) => this.handleSectionError("interfaces", e));
+    const getEndpointHealth = async () => {
+      try {
+        const response = await queryAstarteHealth();
+        if (!ignore) {
+          setStatus("ok");
+        }
 
-    this.astarte.getTriggerNames()
-      .then(this.updateTriggerNames)
-      .catch((e) => this.handleSectionError("triggers", e));
+      } catch (err) {
+        setStatus("err");
+      }
+    };
 
-    this.astarte.getRealmManagementHealth()
-      .then(() => { this.updateStatus("realmManagementStatus", "ok") })
-      .catch(() => { this.updateStatus("realmManagementStatus", "err") });
+    getEndpointHealth();
+    return (() => { ignore = true; });
+  }, [queryAstarteHealth]);
 
-    this.astarte.getAppengineHealth()
-      .then(() => { this.updateStatus("appengineStatus", "ok") })
-      .catch(() => { this.updateStatus("appengineStatus", "err") });
+  return status;
+}
 
-    this.astarte.getPairingHealth()
-      .then(() => { this.updateStatus("pairingStatus", "ok") })
-      .catch(() => { this.updateStatus("pairingStatus", "err") });
+const useAstarteAPI = (fetchData) => {
```

Comment:
This hook already generalizes the pattern used in `useAstarteHealth`. You could abstract a single `useRequest` and get rid of the duplication